File: 572-subtree-of-another-tree/subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isIdentical(TreeNode root,TreeNode subRoot)
    {
        if(root==null && subRoot==null)
        {
            return true;
        }
        else if(root==null || subRoot==null)
        {
            return false;
        }
        
        if(root.val!=subRoot.val)
        {
            return false;
        }
        
       
        
        return isIdentical(root.left,subRoot.left) && isIdentical(root.right,subRoot.right);
        
    }
    public boolean fun(TreeNode root,TreeNode subRoot)
    {
        if(root==null)
        {
            return false;
        }
        if(root.val==subRoot.val)
        {
            if(isIdentical(root,subRoot))
            {
                return true;
            }
        }
        return fun(root.left,subRoot) || fun(root.right,subRoot);
    }
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        return fun(root,subRoot);
        
    }
}

File: 572-subtree-of-another-tree/subtree-of-another-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
bool isIdentical(TreeNode*root,TreeNode*subroot){
    if(subroot==NULL && root==NULL)return true;
    if(root==NULL || subroot==NULL)return false;
    if(root->val!=subroot->val){
        return false;
    }
    return isIdentical(root->left,subroot->left) && isIdentical(root->right,subroot->right);
}
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(subRoot==NULL){
            return true;
        }
        if(root==NULL || subRoot==NULL)return false;
        if(root->val==subRoot->val){
            if(isIdentical(root,subRoot))return true;
        }
        return isSubtree(root->left,subRoot) || isSubtree(root->right,subRoot);
        
    }
};